### DISCUSSION TITLE
Just my two cents.

### DISCUSSION THREAD

**Discussion by kdill00** [blakeblackshear/frigate#5508] [General]
First off, thank you Nick for busting your ass on this. I’ve been following closely because I’ve been using a fork to run tensorrt for 7months now. I have a few questions/comments.  It’s working great.  I wanted to ask what the purpose of using both cuda and cuvid for decoding?  When using both those commands it adds a noticeable amount of cpu to each process, extra hwframes also adds some work to the process which was to my understanding was a fix to several year old resolved bug with ffmpeg and nvidia cards?  Wanted to see if you heard something different.  Removing the extra cuvid decode reduces cpu by 75 percent on each decode process, -hwaccel cuda + output format will decode both h264 and h265 and is what nvidia is specifying at the best way. They have all but stopped recommending the h264_cuvid and hevc_cuvid as cuda provides superior quality and buffering in my experience.  Using -threads 1 helps ALOT as well. Nvidia recommends a command that only allows 2 simultaneous connections to the card at a time which they claim will improve perfomance, haven’t tried it but I’m wondering if reducing thread load is doing something similar. Ffmpeg creates a billion tasks with no regard to how many threads it’s using it seems. This saves 35 percent of total ram usage as well. The other thing is telling it to scale to its current resolution before the hwdownload. Using scale_cuda=passthrough=0:format=yuv420p,hwdownload,format=yuv420p  instead of specifying the inputs resolution creates a sort of decode buffer as it’s passes through the frames without acting on them in any way. This was was recommended by ffmpeg and I have found it does help in high throughput situations. Going back to the extra hwframes, when i moved from the fork to 12, I found a ton of excess frames being acted on, especially with out -vsync 0 added to hwaccel args. Using vsync 0 dramatically improves performance in my setup as it stops the card from dropping frames. Nvidia admits without a lot of frames are lost. I can see the issue if someone is using one 15fps stream to record then the same stream at 5’fps to detect as vsync will override and match camera fps regardless of -r setting, but outside of that it really hurts to not have that in there. The next question or comment is on the timeout setting on input args.  This setting has always seemed to nerf quality of image quite noticeably over the years I’ve tried it. I think it was default 100000 in previous versions and is now 500000 in this one?  Removing it all together drastically increases my video quality and improves accuracy of detection.  I’m just curious if I’m missing something. The last comment i would say is rtsp-simple-server seems to be superior in how it handles the video when proxying it. I was reading somewhere it uses some custom code to align timestamps and smooth the video. With 30 cameras, I had lots of performance issues with go2rtc. It works it just doesn’t seem to correct and smooth video as well. just to clarify, I am probably the extreme end of the situation, my setup runs on an 88 core server with an nvidia a30 and is almost always maxed out due to hundreds of people infront of the cameras. I’m always tweeting to try and get more out of it as it’s walled most times anyways at 500 fps or so on yolov4. This is just my blog post as im curious if you had other reasoning for these commands, that I am not aware of. Everything else is great and it still works fine with default setup. Also in my setup the hardware is almost always maxed out so I spend time a lot of time tweaking and monitoring when I can and have found these things and I’d like to see what you know. This post is the culmination of months of fiddling and experimenting to get the most out of this card and system i have and bringing that fiddling to 12 with me. Any insight you have on how you came to those presets is welcome

**Comment by kdill00** [comment]
, -hwaccel cuda only without the output format is the same as using cuvid. No hwdownload needed and similar resource usage.

**Comment by NickM-27** [comment]
I am sure there is more room for improvement, but we've gotten a lot of remarks (and I have seen myself) that CPU usage is down with 0.12. 

There is a similar discussion here https://github.com/blakeblackshear/frigate/issues/5459 and the general idea is that in a future version we want to split the record and detect roles into separate processes even if they are the same stream to that more aggressive args can be used. Something for the future but not for 0.12

> I wanted to ask what the purpose of using both cuda and cuvid for decoding? When using both those commands it adds a noticeable amount of cpu to each process

I just tested and it does seem that it is not necessary to have the h264_cuvid arg, however I have identical cpu usage with or without it

> Removing the extra cuvid decode reduces cpu by 75 percent on each decode proces

I am not seeing that at all. My cpu usage is 3.7% for a core per stream and it remains that with or without the arg being there.

> The other thing is telling it to scale to its current resolution before the hwdownload. Using scale_cuda=passthrough=0:format=yuv420p,hwdownload,format=yuv420p instead of specifying the inputs resolution creates a sort of decode buffer as it’s passes through the frames without acting on them in any way

sure, but we are providing this so that the camera stream can be scaled using the GPU in case a user wants to use the main stream downscaled in the case that their sub stream is too low resolution. 

> the next question or comment is on the timeout setting on input args. This setting has always seemed to nerf quality of image quite noticeably over the years I’ve tried it. I think it was default 100000 in previous versions and is now 500000 in this one? Removing it all together drastically increases my video quality and improves accuracy of detection. I’m just curious if I’m missing something.

I don't think that makes sense, I don't think that arg should affect image quality at all as it is just based on the TCP socket sending data. It is also set to 5 seconds (unchanged from previous versions of frigate)

https://ffmpeg.org/ffmpeg-protocols.html#toc-Demuxer

**Reply by kdill00** [reply]
I guess with an 88 core server i go from 10 percent of 1 core being used with to 2 or 3.  Its pretty negligible, and I know its a niche case. but when you have 31 cameras at 1280p, a few percentage points will add up fast.  The improvement in quality when removing the socket timeout is similar to when I have tried to use UDP.  The video is less blurry. Maybe better quality is not the best way to describe it.  The image is a bit sharper.   This is all h265 as well.  It could be some kind of other issue with the server.  Its doing most of what i need it to do so I just tweak here and there, I am going to upgrade the server its self soon as its a little older supermicro that cant use some new generation CPUs and what not.  The adventure continues

**Comment by kdill00** [comment]
![956FFD30-DA11-4311-9B02-92327C290C8A](https://user-images.githubusercontent.com/76188629/219270773-d4810fc3-e25e-4a51-8f67-893fc4385801.jpeg)

Bottom is h264 cuvid and default decode and record commands of the past. Top is cuda with passthrough.  It adds together to be about 11 or so percent when they are both in the args.  I still haven’t really seen too much difference between the two except hwaccel cuda will consume way more gpu memory than the other without the threads command especially. 31 cameras using cuda uses about 10gb of gpu memory and with cuvid it’s about 6gb.

**Comment by kdill00** [comment]
@NickM-27 saw you added the threads thing, as a last note on this, using the docker command —cpu-set 0-6 or whatever the least number of cores plus a little headspace you think you can get away seems benefit, probably in cpu cache hits as it decreases my latency in detection a little bit, and seems to smooth play back issues.  This is a soft limit I think and docker will still give resources if it ever gets to the point of exhausting. ffmpeg seems to really benefit from having limits imposed on it whereever possible as I have seen a lot of people mention it tries to use as much as possible no matter what from my scraping around to improve performance.  Nvidia recommends combining at least 4 decode processes to one cpu thread in ffmpeg whenever possible to improve throughout and all that. Again you are a beast thank you for your hard work.
